pallets/price-fetch/src/lib.rs

```diff
@@ -0,0 +1,393 @@
+		Fetchers get(fn fetcher): map hasher(identity) Vec<u8> => Fetcher<T::BlockNumber>;
+
+		///Map of raw fetched_prices from oracle. Key is hash of symbol e.g hash('ETH')
+		FetchedPrices get(fn fetched_prices): map hasher(identity) Vec<u8> => Vec<FetchedPrice<T::AccountId>>;
+
+		///Map of aggregated prices
+		AvgPrices get(fn avg_price): map hasher(identity) Vec<u8> => (T::Moment, Price, T::AccountId);
+	}
+}
+
+decl_error! {
+	pub enum Error for Module<T: Trait> {
+		//Fetcher for required symbol is already running
+		FetcherAlreadyExist,
+		//start fetcher for unsupported symbol (currency/token, e.g ETH
+		SymbolNotFound,
+
+		FetcherNotFound,
+	}
+}
+
+decl_event!(
+	pub enum Event<T>
+	where
+		Moment = <T as pallet_timestamp::Trait>::Moment,
+		AccountId = <T as frame_system::Trait>::AccountId,
+		Price = Price,
+		Symbol = Symbol,
+	{
+		//New fetcher was initialized
+		NewFetcher(AccountId, Symbol, Moment),
+
+		//New price point was saved from symbol
+		NewPricePoint(AccountId, Symbol, Moment, Price),
+
+		//New avg price was calculated and old fetcher was destroyed
+		NewAvgPrice(AccountId, Symbol, Moment, Price),
+	}
+);
+
+decl_module! {
+	/// A public part of the pallet.
+	pub struct Module<T: Trait> for enum Call where origin: T::Origin {
+
+		type Error = Error<T>;
+
+		fn deposit_event() = default;
+
+		///Start fetching price for 600 blocks
+		//TODO: add fetched duration and symbol
+		#[weight = 0]
+		pub fn start_fetcher(origin) -> DispatchResult {
+			let who = ensure_signed(origin)?;
+
+			ensure!(!<Fetchers<T>>::contains_key(&SYM.to_vec()), Error::<T>::FetcherAlreadyExist);
+
+			//TODO: duration should be param of function
+			let end_at = <system::Module<T>>::block_number() + T::BlockNumber::from(600); //600 blocs is 1hour at 1 block/6s
+			let url = match SYMBOLS.iter().find(|(s, _)| s == SYM) {
+				Some (p) => Ok(p.1),
+				None => Err(Error::<T>::SymbolNotFound)
+			}?;
+
+			let new_fetcher = Fetcher {
+				symbol: SYM.to_vec(),
+				end_fetching_at: end_at,
+				url: url.to_vec()
+			};
+			<Fetchers<T>>::insert(SYM.to_vec(), new_fetcher);
+
+			let now = <pallet_timestamp::Module<T>>::get();
+			Self::deposit_event(RawEvent::NewFetcher(who, SYM.to_vec(), now));
+			Ok(())
+		}
+
+		#[weight = 0]
+		pub fn submit_new_price(origin, price_record: DiaPriceRecord) -> DispatchResult {
+			let who = ensure_signed(origin)?;
+
+			ensure!(!<Fetchers<T>>::contains_key(&price_record.symbol), Error::<T>::FetcherNotFound);
+
+			let new_price = FetchedPrice {
+				price: price_record.price,
+				time: price_record.time,
+				symbol: price_record.symbol.clone(),
+				author: who.clone()
+			};
+
+			Self::add_new_price_to_list(new_price);
+
+			let now = <pallet_timestamp::Module<T>>::get();
+			Self::deposit_event(RawEvent::NewPricePoint(who, price_record.symbol, now, price_record.price));
+
+			Ok(())
+		}
+
+		#[weight = 0]
+		pub fn submit_new_avg_price(origin, symbol: Symbol, avg_price:Price) -> DispatchResult {
```

Comment:
this average is not verified, is that on purpose?

Reply:
Verification/validation of submitted values is in progress